File: backend/src/SasPortal.Application/Common/AdManagement/AdComputerAccountGuard.cs
namespace SasPortal.Application.Common.AdManagement;

public static class AdComputerAccountGuard
{
    public const int DomainControllersPrimaryGroupId = 516;
    public const int ServerTrustAccountFlag = 0x2000;
    public const int PartialSecretsAccountFlag = 0x04000000;

    public const string ProtectedComputerMessage =
        "Bu bilgisayar hesabı üzerinde etkinleştirme veya devre dışı bırakma işlemi yapılamaz.";

    public const string ProtectedComputerWriteOperationMessage =
        "Bu bilgisayar hesabı üzerinde bu işlem yapılamaz.";

    public const string ProtectedComputerDeleteMessage =
        "Bu bilgisayar hesabı silinemez.";

    public const string ProtectedComputerGroupMembershipMessage =
        "Bu bilgisayar hesabında grup üyeliği değiştirilemez.";

    public static bool IsProtectedComputer(
        int? primaryGroupId,
        int? userAccountControl,
        bool? isCriticalSystemObject)
    {
        if (isCriticalSystemObject == true)
        {
            return true;
        }

        if (primaryGroupId == DomainControllersPrimaryGroupId)
        {
            return true;
        }

        if (userAccountControl is null)
        {
            return false;
        }

        var flags = userAccountControl.Value;
        if ((flags & ServerTrustAccountFlag) != 0 && (flags & PartialSecretsAccountFlag) != 0)
        {
            return true;
        }

        return false;
    }
}

File: frontend/src/features/ad-management/ad-computer-account-guard.ts
const DOMAIN_CONTROLLERS_PRIMARY_GROUP_ID = 516;
const SERVER_TRUST_ACCOUNT_FLAG = 0x2000;
const PARTIAL_SECRETS_ACCOUNT_FLAG = 0x04000000;

type ComputerAccountGuardSource = {
  primaryGroupId?: number | null;
  userAccountControl?: number | null;
};

export function isAdComputerAccountOperationRestricted(
  computer: ComputerAccountGuardSource,
): boolean {
  if (computer.primaryGroupId === DOMAIN_CONTROLLERS_PRIMARY_GROUP_ID) {
    return true;
  }

  if (computer.userAccountControl == null) {
    return false;
  }

  const flags = computer.userAccountControl;
  return (flags & SERVER_TRUST_ACCOUNT_FLAG) !== 0
    && (flags & PARTIAL_SECRETS_ACCOUNT_FLAG) !== 0;
}
